File: week1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number = get_long("Number: ");
    long n = number;
    int sum = 0;
    bool b;
    while (n > 0)
    {
        int last = n % 10;
        if (b)
        {
            int x2 = last * 2;
            while (x2 > 0)
            {
                int digit = x2 % 10;
                sum += digit;
                x2 /= 10;
            }
            b = false;
        }
        else
        {
            sum += last;
            b = true;
        }
        n /= 10;
    }
    if (sum % 10 == 0)
    {
        if ((number >= 340000000000000 && number < 350000000000000) || (number >= 370000000000000 && number < 380000000000000))
        {
            printf("AMEX\n");
        }
        else if (number >= 5100000000000000 && number < 5600000000000000)
        {
            printf("MASTERCARD\n");
        }
        else if ((number >= 4000000000000 && number < 5000000000000) || (number >= 4000000000000000 && number < 5000000000000000))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: pset6/credit.py
import sys


def main():

    # get number as input from user
    while True:
        number = input("Number: ")
        try:
            int(number)
            break
        except:
            continue

    # Luhn's algorithm
    total = 0

    # starting from the 2nd to last digit, multiply every other digit by 2
    for product in map(lambda x: int(x) * 2, number[-2::-2]):

        # sum the digits of the products
        for digit in str(product):
            total += int(digit)

    # sum the digits not multiplied by 2
    for integer in map(lambda x: int(x), number[::-2]):
        total += integer

    # if total mod 10 is congruent to 0, the number is valid
    if total % 10 == 0:

        # check card type
        if len(number) in (13, 16) and number.startswith('4'):
            print("VISA")
        elif len(number) == 16 and filter(lambda x: number.startswith(str(x)), range(51, 55)):
            print("MASTERCARD")
        elif len(number) == 15 and filter(lambda x: number.startswith(x), ['34', '37']):
            print("AMEX")
        else:
            print("INVALID")

    else:
        print("INVALID")


if __name__ == '__main__':
    main()
